src/main/host/descriptor/descriptor_table.rs

```diff
@@ -0,0 +1,204 @@
+        let mut descriptor = self.descriptors.remove(&idx).unwrap();
+        self.available_indices.push(idx);
+        descriptor.set_handle(0);
+        descriptor
+    }
+
+    /// Get the descriptor at `idx`, if any.
+    pub fn get(&self, idx: u32) -> Option<&CompatDescriptor> {
+        self.descriptors.get(&idx)
+    }
+
+    /// Insert a descriptor at `index`. If a descriptor is already present at
+    /// that index, it is unregistered from that index and returned.
+    pub fn set(&mut self, index: u32, descriptor: CompatDescriptor) -> Option<CompatDescriptor> {
+        let mut descriptor = descriptor;
+        super::export::compatdescriptor_setHandle(
+            &mut descriptor as *mut CompatDescriptor,
+            index.try_into().unwrap(),
+        );
+        if let Some(mut prev) = self.descriptors.insert(index, descriptor) {
+            trace!("Overwriting index {}", index);
+            super::export::compatdescriptor_setHandle(&mut prev as *mut CompatDescriptor, 0);
+            Some(prev)
+        } else {
+            trace!("Setting to unused index {}", index);
+            None
+        }
+    }
+
+    /// This is a helper function that handles some corner cases where some
+    /// descriptors are linked to each other and we must remove that link in
+    /// order to ensure that the reference count reaches zero and they are properly
+    /// freed. Otherwise the circular reference will prevent the free operation.
+    /// TODO: remove this once the TCP layer is better designed.
+    pub fn shutdown_helper(&mut self) {
+        for (_idx, descriptor) in self.descriptors.iter() {
+            match descriptor {
+                CompatDescriptor::New(_) => continue,
+                CompatDescriptor::Legacy(d) => unsafe {
+                    cshadow::descriptor_shutdownHelper(d.ptr())
+                },
+            };
+        }
+    }
+}
+
+mod export {
+    use super::*;
+    use libc::c_int;
+
+    /// Create an object that can be used to store all descriptors created by a
+    /// process. When the table is no longer required, use descriptortable_free
+    /// to release the reference.
+    #[no_mangle]
+    pub unsafe extern "C" fn descriptortable_new() -> *mut DescriptorTable {
+        Box::into_raw(Box::new(DescriptorTable::new()))
+    }
+
+    /// Free the table.
+    #[no_mangle]
+    pub unsafe extern "C" fn descriptortable_free(table: *mut DescriptorTable) {
+        unsafe { Box::from_raw(notnull_mut_debug(table)) };
+    }
+
+    /// Store a descriptor object for later reference at the next available index
+    /// in the table. The chosen table index is stored in the descriptor object and
+    /// returned. The descriptor is guaranteed to be stored successfully.
+    ///
+    /// NOTE: that this consumes a reference to the descriptor, so if you are storing
+    /// it outside of the descriptor table you will need to ref it after calling
+    /// this function.
```

Comment:
These comments from `descriptor_table.h` are out of date. The `descriptor object` doesn't generally store the table index anymore (only the legacy descriptors do).

Also, the descriptor table consumes the descriptor, so it should not be stored or used after this function call.

Reply:
Done